sw/device/lib/dif/README.md

```diff
@@ -12,3 +12,54 @@ libraries.
 There is one DIF library per hardware IP, and each one contains the DIFs
 required to actuate all of the specification-required functionality of the
 hardware they are written for.
+
+## DIF Style Guide
+
+DIFs are very low-level software, so they have a more rigorous coding style than
+other parts of the codebase.
+
+DIFs should follow the [OpenTitan C/C++ style
+guide](https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) where it does not
+contradict with the guidelines below.
+
+The guidelines below apply to writing DIFs, and code should be written in a
+similar style to the existing DIF libraries in this directory.
+
+### DIF Library Guidance
+
+* DIF libraries must be written in C.
+* DIF libraries can only depend on the headers and libraries in the
+  `sw/device/lib/base` directory.
+* DIF libraries must not depend on other DIF libraries. Exercising DIF
+  functionality may require an environment set up using another DIF library, but
+  DIFs must not call DIFs in other DIF libraries.
+
+### DIF Guidance
+
+* DIF declarations must match their definitions exactly.
+  * Scalar arguments should not be marked `const` in DIF signatures.
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can error or produce a value, it should return an enum value, and use an
+  out-parameter for returning the value.
+
+* DIFs must check their arguments against preconditions. DIFs must "early
+  return" with an error code if their arguments fail the following checks:
+  * DIFs must ensure their pointer arguments are non-null, unless that pointer
+    is for an optional out-parameter.
+  * DIFs must ensure, if they only accept a subset of an enum, that the argument
+    is within that subset. However, DIFs may assume that the argument is one of
+    the enum variants.
+  * DIFs must not have side-effects before an "early return". Side-effects
```

Comment:
Given that there are naturally going to be error (early return cases) other than precondition checks that would have to involve side effects, maybe rephrase. "DIFs must now have side effects before an "early return" due to a precondition check failure."

Reply:
I intended "early return" to be a clearer explanation of the "guard" construct mentioned in the document. It is not clearer, as "early return" is a slightly different concept, so I will rephrase this in terms of guard statements.